File: mobile/android/app/src/main/kotlin/com/opensourceortho/lite/EngineModels.kt
package com.opensourceortho.lite

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/**
 * Serializable mirrors of the mobile-facing subset of the engine contract
 * (../../API_CONTRACT.md). Only the fields the lite UI renders are modeled; the
 * JSON parser is configured with ignoreUnknownKeys so the engine can evolve.
 */

/** Request body for `POST /api/generate-plan`. Lite defaults keep everything on
 *  the engine host: provider = "local", shareAcknowledged = false. */
@Serializable
data class GeneratePlanRequest(
    val plan: JsonElement,
    @SerialName("acknowledge_educational") val acknowledgeEducational: Boolean = true,
    val provider: String = "local",
    @SerialName("share_acknowledged") val shareAcknowledged: Boolean = false,
    val notes: String? = null,
)

@Serializable
data class PipelineStep(
    val name: String,
    val status: String = "ok",
    val detail: String = "",
)

@Serializable
data class Correctness(val verdict: String)

@Serializable
data class Timeline(
    @SerialName("stage_count") val stageCount: Int,
    @SerialName("wear_interval_days") val wearIntervalDays: Int,
    @SerialName("projected_duration_days") val projectedDurationDays: Int,
    @SerialName("projected_duration_weeks") val projectedDurationWeeks: Double,
    val caveat: String,
)

/** Subset of the `POST /api/generate-plan` success body the lite UI renders. */
@Serializable
data class GeneratePlanResponse(
    val ok: Boolean = false,
    val errors: List<String>? = null,
    val source: String? = null,
    val warnings: List<String>? = null,
    val steps: List<PipelineStep>? = null,
    val correctness: Correctness? = null,
    @SerialName("stage_count") val stageCount: Int? = null,
    val timeline: Timeline? = null,
    val caveat: String? = null,
    // The full generated plan stays opaque here; the progression screen reads
    // `plan.stages` lazily when the renderer is built.
    val plan: JsonElement? = null,
)

@Serializable
data class StoredCaseReview(
    val schema: String,
    val kind: String,
    @SerialName("case_id") val caseId: String,
    @SerialName("plan_id") val planId: String,
    val title: String,
    @SerialName("review_tier") val reviewTier: ReviewTierSummary,
    @SerialName("unresolved_data_gaps") val unresolvedDataGaps: List<ReviewDataGap>,
    @SerialName("cbct_status") val cbctStatus: String,
    @SerialName("root_bone_review") val rootBoneReview: RootBoneReviewSummary,
    @SerialName("findings_summary") val findingsSummary: FindingsSummary,
    val editable: EditLockSummary,
    val handoff: CaseHandoffSummary,
    @SerialName("plan_sha256") val planSha256: String,
    @SerialName("review_sha256") val reviewSha256: String,
) {
    val isImportableStoredReview: Boolean
        get() = schema == "orthoplan-case-review-v1" &&
            kind == "stored-review" &&
            !editable.inMobile &&
            editable.requiresBrowserEngine

    val mobileSummary: String
        get() = "${reviewTier.label} - ${unresolvedDataGaps.size} unresolved data gaps"
}

@Serializable
data class ReviewTierSummary(
    val tier: String,
    val rank: Int,
    val label: String,
    val summary: String,
    @SerialName("root_bone_aware") val rootBoneAware: Boolean,
)

@Serializable
data class ReviewDataGap(
    val domain: String,
    val reason: String,
)

@Serializable
data class RootBoneReviewSummary(val verdict: String)

@Serializable
data class FindingsSummary(
    val total: Int,
    @SerialName("by_severity") val bySeverity: Map<String, Int> = emptyMap(),
)

@Serializable
data class EditLockSummary(
    @SerialName("in_mobile") val inMobile: Boolean,
    @SerialName("requires_browser_engine") val requiresBrowserEngine: Boolean,
    val reason: String,
)

@Serializable
data class CaseHandoffSummary(
    @SerialName("case_id") val caseId: String,
    @SerialName("open_url") val openUrl: String? = null,
    @SerialName("deep_link") val deepLink: String,
    @SerialName("qr_payload") val qrPayload: String,
)

File: mobile/ios/Sources/OpenSourceOrthoLiteKit/EngineModels.swift
import Foundation

/// Codable mirrors of the mobile-facing subset of the engine contract
/// (../../API_CONTRACT.md). Only the fields the lite UI renders are modeled;
/// unknown fields are ignored so the engine can evolve without breaking decode.

/// Request body for `POST /api/generate-plan`. Lite defaults keep everything on
/// the engine host: `provider = "local"`, `shareAcknowledged = false`.
public struct GeneratePlanRequest: Codable, Sendable {
    public var plan: [String: AnyCodable]
    public var acknowledgeEducational: Bool
    public var provider: String
    public var shareAcknowledged: Bool
    public var notes: String?

    public init(
        plan: [String: AnyCodable],
        acknowledgeEducational: Bool = true,
        provider: String = "local",
        shareAcknowledged: Bool = false,
        notes: String? = nil
    ) {
        self.plan = plan
        self.acknowledgeEducational = acknowledgeEducational
        self.provider = provider
        self.shareAcknowledged = shareAcknowledged
        self.notes = notes
    }

    enum CodingKeys: String, CodingKey {
        case plan
        case acknowledgeEducational = "acknowledge_educational"
        case provider
        case shareAcknowledged = "share_acknowledged"
        case notes
    }
}

public struct PipelineStep: Codable, Sendable, Identifiable {
    public var name: String
    public var status: String
    public var detail: String
    public var id: String { name }

    public init(name: String, status: String = "ok", detail: String = "") {
        self.name = name
        self.status = status
        self.detail = detail
    }
}

public struct Correctness: Codable, Sendable {
    public var verdict: String
    public init(verdict: String) { self.verdict = verdict }
}

public struct Timeline: Codable, Sendable {
    public var stageCount: Int
    public var wearIntervalDays: Int
    public var projectedDurationDays: Int
    public var projectedDurationWeeks: Double
    public var caveat: String

    public init(
        stageCount: Int,
        wearIntervalDays: Int,
        projectedDurationDays: Int,
        projectedDurationWeeks: Double,
        caveat: String
    ) {
        self.stageCount = stageCount
        self.wearIntervalDays = wearIntervalDays
        self.projectedDurationDays = projectedDurationDays
        self.projectedDurationWeeks = projectedDurationWeeks
        self.caveat = caveat
    }

    enum CodingKeys: String, CodingKey {
        case stageCount = "stage_count"
        case wearIntervalDays = "wear_interval_days"
        case projectedDurationDays = "projected_duration_days"
        case projectedDurationWeeks = "projected_duration_weeks"
        case caveat
    }
}

/// Subset of the `POST /api/generate-plan` success body the lite UI renders.
public struct GeneratePlanResponse: Codable, Sendable {
    public var ok: Bool
    public var errors: [String]?
    public var source: String?
    public var warnings: [String]?
    public var steps: [PipelineStep]?
    public var correctness: Correctness?
    public var stageCount: Int?
    public var timeline: Timeline?
    public var caveat: String?
    /// The full generated plan; stages drive progression rendering. Kept opaque
    /// here - the progression view reads `plan.stages` lazily.
    public var plan: AnyCodable?

    public init(
        ok: Bool = false,
        errors: [String]? = nil,
        source: String? = nil,
        warnings: [String]? = nil,
        steps: [PipelineStep]? = nil,
        correctness: Correctness? = nil,
        stageCount: Int? = nil,
        timeline: Timeline? = nil,
        caveat: String? = nil,
        plan: AnyCodable? = nil
    ) {
        self.ok = ok
        self.errors = errors
        self.source = source
        self.warnings = warnings
        self.steps = steps
        self.correctness = correctness
        self.stageCount = stageCount
        self.timeline = timeline
        self.caveat = caveat
        self.plan = plan
    }

    enum CodingKeys: String, CodingKey {
        case ok, errors, source, warnings, steps, correctness, timeline, caveat, plan
        case stageCount = "stage_count"
    }
}
